File: trunk/AI/RAI/python3-cgdk/RemoteProcessClient.py
import _socket
import struct
from model.Bonus import Bonus
from model.BonusType import BonusType
from model.Obstacle import Obstacle
from model.Player import Player
from model.PlayerContext import PlayerContext
from model.Shell import Shell
from model.ShellType import ShellType
from model.Tank import Tank
from model.TankType import TankType
from model.World import World

class RemoteProcessClient:
    LITTLE_ENDIAN_BYTE_ORDER = True

    SIGNED_BYTE_FORMAT_STRING = '<b' if LITTLE_ENDIAN_BYTE_ORDER else '>b'
    INTEGER_FORMAT_STRING = '<i' if LITTLE_ENDIAN_BYTE_ORDER else '>i'
    LONG_FORMAT_STRING = '<q' if LITTLE_ENDIAN_BYTE_ORDER else '>q'
    DOUBLE_FORMAT_STRING = '<d' if LITTLE_ENDIAN_BYTE_ORDER else '>d'

    SIGNED_BYTE_SIZE_BYTES = 1
    INTEGER_SIZE_BYTES = 4
    LONG_SIZE_BYTES = 8
    DOUBLE_SIZE_BYTES = 8

    def __init__(self, host, port):
        self.socket = _socket.socket()
        self.socket.connect((host, port))

    def write_token(self, token):
        self.write_enum(RemoteProcessClient.MessageType.AUTHENTICATION_TOKEN)
        self.write_string(token)

    def read_team_size(self):
        message_type = self.read_enum(RemoteProcessClient.MessageType)
        self.ensure_message_type(message_type, RemoteProcessClient.MessageType.TEAM_SIZE)
        return self.read_int()

    def write_selected_tanks(self, tank_types):
        self.write_enum(RemoteProcessClient.MessageType.TANK_TYPES)

        if tank_types is None:
            self.write_int(-1)
        else:
            self.write_int(tank_types.__len__())

            for type in tank_types:
                self.write_enum(type)

    def read_player_context(self):
        message_type = self.read_enum(RemoteProcessClient.MessageType)
        if message_type == RemoteProcessClient.MessageType.GAME_OVER:
            return None

        self.ensure_message_type(message_type, RemoteProcessClient.MessageType.PLAYER_CONTEXT)
        return PlayerContext(self.read_tanks(), self.read_world()) if self.read_boolean() else None

    def write_moves(self, moves):
        self.write_enum(RemoteProcessClient.MessageType.MOVES)

        if moves is None:
            self.write_int(-1)
        else:
            self.write_int(moves.__len__())

            for move in moves:
                if move is None:
                    self.write_boolean(False)
                else:
                    self.write_boolean(True)

                    self.write_double(move.left_track_power)
                    self.write_double(move.right_track_power)
                    self.write_double(move.turret_turn)
                    self.write_enum(move.fire_type)

    def close(self):
        self.socket.close()

    def read_world(self):
        if not self.read_boolean():
            return None

        return World(
            self.read_int(), self.read_double(), self.read_double(), self.read_players(),
            self.read_obstacles(), self.read_tanks(), self.read_shells(), self.read_bonuses()
        )

    def read_players(self):
        player_count = self.read_int()
        if player_count < 0:
            return None

        players = []

        for player_index in range(player_count):
            if self.read_boolean():
                player = Player(self.read_string(), self.read_int(), self.read_boolean())
                players.append(player)
            else:
                players.append(None)

        return players

    def read_obstacles(self):
        obstacle_count = self.read_int()
        if obstacle_count < 0:
            return None

        obstacles = []

        for obstacle_index in range(obstacle_count):
            if self.read_boolean():
                obstacle = Obstacle(
                    self.read_long(), self.read_double(), self.read_double(),
                    self.read_double(), self.read_double()
                )
                obstacles.append(obstacle)
            else:
                obstacles.append(None)

        return obstacles

    def read_tanks(self):
        tank_count = self.read_int()
        if tank_count < 0:
            return None

        tanks = []

        for tank_index in range(tank_count):
            if self.read_boolean():
                tank = Tank(
                    self.read_long(), self.read_string(), self.read_int(),
                    self.read_double(), self.read_double(), self.read_double(), self.read_double(),
                    self.read_double(), self.read_double(), self.read_double(),
                    self.read_int(), self.read_int(), self.read_int(), self.read_int(),
                    self.read_int(), self.read_boolean(), self.read_enum(TankType)
                )
                tanks.append(tank)
            else:
                tanks.append(None)

        return tanks

    def read_shells(self):
        shell_count = self.read_int()
        if shell_count < 0:
            return None

        shells = []

        for shell_index in range(shell_count):
            if self.read_boolean():
                shell = Shell(
                    self.read_long(), self.read_string(), self.read_double(), self.read_double(),
                    self.read_double(), self.read_double(), self.read_double(), self.read_double(),
                    self.read_double(), self.read_double(), self.read_enum(ShellType)
                )
                shells.append(shell)
            else:
                shells.append(None)

        return shells

    def read_bonuses(self):
        bonus_count = self.read_int()
        if bonus_count < 0:
            return None

        bonuses = []

        for bonus_index in range(bonus_count):
            if self.read_boolean():
                bonus = Bonus(
                    self.read_long(), self.read_double(), self.read_double(),
                    self.read_double(), self.read_double(), self.read_enum(BonusType)
                )
                bonuses.append(bonus)
            else:
                bonuses.append(None)

        return bonuses

    def ensure_message_type(self, actual_type, expected_type):
        if actual_type != expected_type:
            raise ValueError("Received wrong message [actual=%s, expected=%s]" % (actual_type, expected_type))

    def read_enum(self, enum_class):
        bytes = self.read_bytes(RemoteProcessClient.SIGNED_BYTE_SIZE_BYTES)
        value = struct.unpack(RemoteProcessClient.SIGNED_BYTE_FORMAT_STRING, bytes)[0]

        for enum_key, enum_value in enum_class.__dict__.items():
            if not str(enum_key).startswith("__") and value == enum_value:
                return enum_value

        return None

    def write_enum(self, value):
        self.write_bytes(struct.pack(RemoteProcessClient.SIGNED_BYTE_FORMAT_STRING, -1 if value is None else value))

    def read_string(self):
        length = self.read_int()
        if length == -1:
            return None

        bytes = self.read_bytes(length)
        return bytes.decode()

    def write_string(self, value):
        if value is None:
            self.write_int(-1)
            return

        bytes = value.encode()

        self.write_int(len(bytes))
        self.write_bytes(bytes)

    def read_boolean(self):
        bytes = self.read_bytes(RemoteProcessClient.SIGNED_BYTE_SIZE_BYTES)
        return struct.unpack(RemoteProcessClient.SIGNED_BYTE_FORMAT_STRING, bytes)[0] == 1

    def write_boolean(self, value):
        self.write_bytes(struct.pack(RemoteProcessClient.SIGNED_BYTE_FORMAT_STRING, 1 if value else 0))

    def read_int(self):
        bytes = self.read_bytes(RemoteProcessClient.INTEGER_SIZE_BYTES)
        return struct.unpack(RemoteProcessClient.INTEGER_FORMAT_STRING, bytes)[0]

    def write_int(self, value):
        self.write_bytes(struct.pack(RemoteProcessClient.INTEGER_FORMAT_STRING, value))

    def read_long(self):
        bytes = self.read_bytes(RemoteProcessClient.LONG_SIZE_BYTES)
        return struct.unpack(RemoteProcessClient.LONG_FORMAT_STRING, bytes)[0]

    def write_long(self, value):
        self.write_bytes(struct.pack(RemoteProcessClient.LONG_FORMAT_STRING, value))

    def read_double(self):
        bytes = self.read_bytes(RemoteProcessClient.DOUBLE_SIZE_BYTES)
        return struct.unpack(RemoteProcessClient.DOUBLE_FORMAT_STRING, bytes)[0]

    def write_double(self, value):
        self.write_bytes(struct.pack(RemoteProcessClient.DOUBLE_FORMAT_STRING, value))

    def read_bytes(self, byte_count):
        data = bytes()

        while len(data) < byte_count:
            chunk = self.socket.recv(byte_count - len(data))

            if not len(chunk):
                raise IOError("Can't read %s bytes from input stream." % str(byte_count))

            data += chunk

        return data

    def write_bytes(self, bytes):
        self.socket.sendall(bytes)

    class MessageType:
        UNKNOWN = 0
        GAME_OVER = 1
        AUTHENTICATION_TOKEN = 2
        TEAM_SIZE = 3
        TANK_TYPES = 4
        PLAYER_CONTEXT = 5
        MOVES = 6

File: AI/RAI/java-cgdk/src/main/java/RemoteProcessClient.java
import model.*;

import java.io.*;
import java.net.Socket;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;

public final class RemoteProcessClient implements Closeable {
    private static final int BUFFER_SIZE_BYTES = 1 << 20;
    private static final ByteOrder PROTOCOL_BYTE_ORDER = ByteOrder.LITTLE_ENDIAN;
    private static final int INTEGER_SIZE_BYTES = Integer.SIZE / Byte.SIZE;
    private static final int LONG_SIZE_BYTES = Long.SIZE / Byte.SIZE;

    private final Socket socket;
    private final InputStream inputStream;
    private final OutputStream outputStream;
    private final ByteArrayOutputStream outputStreamBuffer;

    public RemoteProcessClient(String host, int port) throws IOException {
        socket = new Socket(host, port);
        socket.setSendBufferSize(BUFFER_SIZE_BYTES);
        socket.setReceiveBufferSize(BUFFER_SIZE_BYTES);

        inputStream = socket.getInputStream();
        outputStream = socket.getOutputStream();
        outputStreamBuffer = new ByteArrayOutputStream(BUFFER_SIZE_BYTES);
    }

    public void writeToken(String token) throws IOException {
        writeEnum(MessageType.AUTHENTICATION_TOKEN);
        writeString(token);
        flush();
    }

    public int readTeamSize() throws IOException {
        ensureMessageType(readEnum(MessageType.class), MessageType.TEAM_SIZE);
        return readInt();
    }

    public void writeSelectedTanks(TankType[] tankTypes) throws IOException {
        writeEnum(MessageType.TANK_TYPES);

        if (tankTypes == null) {
            writeInt(-1);
        } else {
            int typeCount = tankTypes.length;
            writeInt(typeCount);

            for (int typeIndex = 0; typeIndex < typeCount; ++typeIndex) {
                writeEnum(tankTypes[typeIndex]);
            }
        }

        flush();
    }

    public PlayerContext readPlayerContext() throws IOException {
        MessageType messageType = readEnum(MessageType.class);
        if (messageType == MessageType.GAME_OVER) {
            return null;
        }

        ensureMessageType(messageType, MessageType.PLAYER_CONTEXT);
        return readBoolean() ? new PlayerContext(readTanks(), readWorld()) : null;
    }

    public void writeMoves(Move[] moves) throws IOException {
        writeEnum(MessageType.MOVES);

        if (moves == null) {
            writeInt(-1);
        } else {
            int moveCount = moves.length;
            writeInt(moveCount);

            for (int moveIndex = 0; moveIndex < moveCount; ++moveIndex) {
                Move move = moves[moveIndex];

                if (move == null) {
                    writeBoolean(false);
                } else {
                    writeBoolean(true);

                    writeDouble(move.getLeftTrackPower());
                    writeDouble(move.getRightTrackPower());
                    writeDouble(move.getTurretTurn());
                    writeEnum(move.getFireType());
                }
            }
        }

        flush();
    }

    @Override
    public void close() throws IOException {
        socket.close();
    }

    private World readWorld() throws IOException {
        if (!readBoolean()) {
            return null;
        }

        return new World(
                readInt(), readDouble(), readDouble(),
                readPlayers(), readObstacles(), readTanks(), readShells(), readBonuses()
        );
    }

    private Player[] readPlayers() throws IOException {
        int playerCount = readInt();
        if (playerCount < 0) {
            return null;
        }

        Player[] players = new Player[playerCount];

        for (int playerIndex = 0; playerIndex < playerCount; ++playerIndex) {
            if (readBoolean()) {
                players[playerIndex] = new Player(readString(), readInt(), readBoolean());
            }
        }

        return players;
    }

    private Obstacle[] readObstacles() throws IOException {
        int obstacleCount = readInt();
        if (obstacleCount < 0) {
            return null;
        }

        Obstacle[] obstacles = new Obstacle[obstacleCount];

        for (int obstacleIndex = 0; obstacleIndex < obstacleCount; ++obstacleIndex) {
            if (readBoolean()) {
                obstacles[obstacleIndex] = new Obstacle(
                        readLong(), readDouble(), readDouble(), readDouble(), readDouble()
                );
            }
        }

        return obstacles;
    }

    private Tank[] readTanks() throws IOException {
        int tankCount = readInt();
        if (tankCount < 0) {
            return null;
        }

        Tank[] tanks = new Tank[tankCount];

        for (int tankIndex = 0; tankIndex < tankCount; ++tankIndex) {
            if (readBoolean()) {
                tanks[tankIndex] = new Tank(
                        readLong(), readString(), readInt(), readDouble(), readDouble(),
                        readDouble(), readDouble(), readDouble(), readDouble(), readDouble(),
                        readInt(), readInt(), readInt(), readInt(), readInt(),
                        readBoolean(), readEnum(TankType.class)
                );
            }
        }

        return tanks;
    }

    private Shell[] readShells() throws IOException {
        int shellCount = readInt();
        if (shellCount < 0) {
            return null;
        }

        Shell[] shells = new Shell[shellCount];

        for (int shellIndex = 0; shellIndex < shellCount; ++shellIndex) {
            if (readBoolean()) {
                shells[shellIndex] = new Shell(
                        readLong(), readString(), readDouble(), readDouble(), readDouble(), readDouble(),
                        readDouble(), readDouble(), readDouble(), readDouble(), readEnum(ShellType.class)
                );
            }
        }

        return shells;
    }

    private Bonus[] readBonuses() throws IOException {
        int bonusCount = readInt();
        if (bonusCount < 0) {
            return null;
        }

        Bonus[] bonuses = new Bonus[bonusCount];

        for (int bonusIndex = 0; bonusIndex < bonusCount; ++bonusIndex) {
            if (readBoolean()) {
                bonuses[bonusIndex] = new Bonus(
                        readLong(), readDouble(), readDouble(), readDouble(), readDouble(), readEnum(BonusType.class)
                );
            }
        }

        return bonuses;
    }

    private static void ensureMessageType(MessageType actualType, MessageType expectedType) {
        if (actualType != expectedType) {
            throw new IllegalArgumentException(String.format(
                    "Received wrong message [actual=%s, expected=%s].", actualType, expectedType
            ));
        }
    }

    private <E extends Enum> E readEnum(Class<E> enumClass) throws IOException {
        byte ordinal = readBytes(1)[0];

        E[] values = enumClass.getEnumConstants();
        int valueCount = values.length;

        for (int valueIndex = 0; valueIndex < valueCount; ++valueIndex) {
            E value = values[valueIndex];
            if (value.ordinal() == ordinal) {
                return value;
            }
        }

        return null;
    }

    private <E extends Enum> void writeEnum(E value) throws IOException {
        writeBytes(new byte[]{(byte) (value == null ? -1 : value.ordinal())});
    }

    private String readString() throws IOException {
        int length = readInt();
        if (length == -1) {
            return null;
        }

        try {
            return new String(readBytes(length), "UTF-8");
        } catch (UnsupportedEncodingException e) {
            throw new IllegalArgumentException("UTF-8 is unsupported.", e);
        }
    }

    private void writeString(String value) throws IOException {
        if (value == null) {
            writeInt(-1);
            return;
        }

        byte[] bytes;
        try {
            bytes = value.getBytes("UTF-8");
        } catch (UnsupportedEncodingException e) {
            throw new IllegalArgumentException("UTF-8 is unsupported.", e);
        }

        writeInt(bytes.length);
        writeBytes(bytes);
    }

    private boolean readBoolean() throws IOException {
        return readBytes(1)[0] == 1;
    }

    private void writeBoolean(boolean value) throws IOException {
        writeBytes(new byte[]{(byte) (value ? 1 : 0)});
    }

    private int readInt() throws IOException {
        return ByteBuffer.wrap(readBytes(INTEGER_SIZE_BYTES)).order(PROTOCOL_BYTE_ORDER).getInt();
    }

    private void writeInt(int value) throws IOException {
        writeBytes(ByteBuffer.allocate(INTEGER_SIZE_BYTES).order(PROTOCOL_BYTE_ORDER).putInt(value).array());
    }

    private long readLong() throws IOException {
        return ByteBuffer.wrap(readBytes(LONG_SIZE_BYTES)).order(PROTOCOL_BYTE_ORDER).getLong();
    }

    private void writeLong(long value) throws IOException {
        writeBytes(ByteBuffer.allocate(LONG_SIZE_BYTES).order(PROTOCOL_BYTE_ORDER).putLong(value).array());
    }

    private double readDouble() throws IOException {
        return Double.longBitsToDouble(readLong());
    }

    private void writeDouble(double value) throws IOException {
        writeLong(Double.doubleToLongBits(value));
    }

    private byte[] readBytes(int byteCount) throws IOException {
        byte[] bytes = new byte[byteCount];
        int offset = 0;
        int readByteCount;

        while (offset < byteCount && (readByteCount = inputStream.read(bytes, offset, byteCount - offset)) != -1) {
            offset += readByteCount;
        }

        if (offset != byteCount) {
            throw new IOException(String.format("Can't read %d bytes from input stream.", byteCount));
        }

        return bytes;
    }

    private void writeBytes(byte[] bytes) throws IOException {
        outputStreamBuffer.write(bytes);
    }

    private void flush() throws IOException {
        outputStream.write(outputStreamBuffer.toByteArray());
        outputStreamBuffer.reset();
        outputStream.flush();
    }

    private enum MessageType {
        UNKNOWN,
        GAME_OVER,
        AUTHENTICATION_TOKEN,
        TEAM_SIZE,
        TANK_TYPES,
        PLAYER_CONTEXT,
        MOVES
    }
}
